docs: Update the Bentos doc (#4246)
bentoml/BentoML: docs/source/concepts/bento.rst

===============

> Building Bentos

Reviewer: quick ask: Can you also check everywhere else when we reference this file, you might also need to update its link as well.

Author: Thanks. The files reference this link directly should be ok as I did not change the file name, but I updated all links that reference sections with the Bentos doc. I also did the same for the Models and Services doc.

======

> A :ref:`Bento <reference/core:bentoml.Bento>` is an archive containing all essential components - source

Reviewer: ```suggestion
A :ref:`Bento <reference/core:bentoml.Bento>` is a format containing all essential components - source
```

After built, each Bento is automatically tagged with a unique version. It is also possible to set a specific version using the ``--version`` option,
but this is generally unnecessary. Only use it if your team has a very specific naming convention for deployable artifacts.

.. code-block:: bash

    bentoml build --version 1.0.1

.. note::

    Building a Bento involves importing the ``bentoml.Service`` object and its dependencies.
    Make sure you have all the dependencies installed in the build environment.
    Support for building from a Docker environment is on the roadmap; see :issue:`2495` for details.

Custom build context
^^^^^^^^^^^^^^^^^^^^

For projects that are part of a larger codebase and interact with other local Python
modules or those containing multiple Bentos/Services, it might not be possible to
put all Service definition code and ``bentofile.yaml`` in the project's root directory.

BentoML allows the placement of the Service definition and ``bentofile.yaml`` anywhere in the project directory.
In such scenarios, specify the ``build_ctx`` and ``bentofile`` arguments when running the ``bentoml build`` command.

> * ``build_ctx``: The build context represents the working directory of your Python project. It's the location where the Python interpreter starts,

Reviewer: ```suggestion
* ``build_ctx``: The build context represents the working directory of your Python project. It will be prepended to the PYTHONPATH during build process,
```

  ensuring the correct import of local Python modules. By default, it's set to the current directory where the ``bentoml build`` command is executed.
* ``bentofile``: A YAML configuration file that specifies the :ref:`concepts/bento:Bento build options`. It defaults to the ``bentofile.yaml`` file in the build context.

To customize their values, use the following:

.. code-block:: bash

>     bentoml build -f ./src/my_project_a/bento_fraud_detect.yaml ./src/

Reviewer: ```suggestion
    $ bentoml build -f ./src/my_project_a/bento_fraud_detect.yaml ./src/
```

Structure
^^^^^^^^^

By default, all created Bentos are stored in the BentoML Bento Store, which is essentially a local directory. You can go to a specific Bento directory by running the following command:

.. code-block:: bash

    $ cd $(bentoml get BENTO_TAG -o path)

Inside the directory, you might see different files and sub-directories depending on the configurations in ``bentofile.yaml``. A typical Bento contains the following key sub-directories:

* ``src``: Contains files specified in the :ref:`include <concepts/bento:Files to include>` field of ``bentofile.yaml``. These
  files are relative to user Python code's CWD (current working directory), which makes importing relative modules and file paths inside user code possible.
* ``apis``: Contains API definitions auto-generated from the Service's API specifications.

> * ``env``: Contains environment-related files for Bento initialization. These files are generated based on the :ref:`concepts/bento:Bento build options` specified in ``bentofile.yaml``.

Reviewer: imo we can show the tree graph of this and have comments for specific folder fuctions.

Author: The reason why I did not show this is that the tree graph can be different if users are using different BentoML versions, so I only listed the key subdirectories here. For example, the `models` directory does not exist in the latest BentoML version, right?

Reviewer: ah that makes sense.
